### Ques2: Analyze data3 using linear regression


#### (a) Find the parameters using direct method and iterative methods

In [177]:
df3=pd.read_csv('data3.csv',header=None)
df3

0        1       2        3        4        5        6         7    \
0    1.74920  5.35240  1.9048  3.88020  6.25810 -0.81882  0.15429  5.547400   
1    2.56720  0.62960  6.1277  9.17090  8.20650  1.68810 -1.18850 -2.611500   
2    5.24510 -0.81218  4.8901 -0.61827 -0.14016  2.86660  1.31370  8.395400   
3    0.26811  2.84790  3.3763  5.97190  6.67580 -3.59950  7.57350  0.099107   
4    3.16490  5.86330  8.8108  6.04410 -2.84140  1.55760  7.15950  4.748200   
..       ...      ...     ...      ...      ...      ...      ...       ...   
795 -2.79630  2.14890  8.8594 -0.94036  4.74310  5.13220  2.75250  8.354000   
796 -4.31980 -2.28510  3.3545  5.96220  0.46250  7.91490 -3.72390  0.209850   
797  0.93772 -2.36160 -1.6969  7.40660 -4.56800 -2.91120  2.48260 -4.629400   
798  6.80770  9.59510 -3.2787  6.58890  1.92600  6.13950  6.70350  9.776700   
799  5.97890  3.94710  5.7992 -4.21550 -1.94250 -4.47530 -2.03030 -1.002300   

          8       9    ...     191       192      193       194      195  \
0     5.03480  8.1049  ... -2.1593  -0.51192 -2.56690  -3.60630 -2.01050   
1    10.97300 -3.3615  ...  1.1644   2.45490  7.52300  -3.28330 -0.36977   
2     1.96440  3.2947  ...  4.1030   5.25320  4.43210   9.46530 -0.89669   
3     0.80194 -2.5541  ...  8.8078  -4.26590  4.89840   4.74880  8.08550   
4    -3.39500 -4.5881  ... -4.6686  -3.18060  4.90400   8.73860  1.34630   
..        ...     ...  ...     ...       ...      ...       ...      ...   
795  -4.00710  5.2895  ...  4.9511   5.35940 -0.33292   0.60967  1.55470   
796   6.11430  6.5968  ...  7.3548  -2.32700  2.87490  10.03200 -3.79820   
797   5.00280  6.9308  ...  1.9006  10.70300  2.34180   1.14440  7.68360   
798   0.83240  8.3612  ... -4.5515  10.86100 -1.75380   8.64750  8.94460   
799   1.52780 -2.0802  ...  3.8967   2.33420  6.80870   3.19800  4.49820   

          196       197      198       199      200  
0     1.87000  -3.94740 -3.95660  3.646400  -6.3990  
1    -2.55980   2.06930 -4.78550  2.174800  36.5410  
2     2.97720  -1.92560  0.69556  2.621600  -9.2496  
3     9.65790   2.23120 -1.70540 -1.014300  -9.6137  
4     9.25870   0.60414 -1.87180  5.360400   6.0652  
..        ...       ...      ...       ...      ...  
795   3.29050  10.13000  3.02670 -4.400600  30.2870  
796   3.35680   1.29950 -1.90100 -0.004131  -1.2650  
797  10.06300  -0.56802 -0.47637  1.081300 -36.0920  
798   0.91756  -1.02130 -4.23760 -2.314900 -50.1770  
799  -4.50170  -1.99050  9.32260 -0.951180 -47.7450  

[800 rows x 201 columns]

In [131]:
direct_method(df3)

array([ 0.09310757, -2.10682807, -0.19724205, -0.35373966,  0.2854454 ,
        0.09588752,  0.54671604,  0.692285  , -0.18641958, -0.21134044,
       -0.6712553 , -0.192262  ,  0.44240955,  1.15903847, -0.6304527 ,
        2.14520606,  0.01552825,  0.1327635 ,  0.02277371, -0.75102492,
        0.5973356 , -0.34103414, -0.03478571,  0.08207365, -0.29841223,
        0.09104487, -1.14462358,  0.10153202,  0.03141346,  0.03070883,
        0.34441512,  0.9178849 , -0.10846267,  0.62473776,  0.05703968,
        0.56453551, -0.17474632,  0.39545945,  0.02829591,  0.26112935,
       -0.12912843,  0.37818224,  0.23465376,  0.47688475,  0.56524095,
        0.06658826,  0.18355772,  0.18277969,  0.15925774, -0.39577749,
        0.27486617, -0.20190807,  0.05936441,  2.24552472, -0.58004898,
       -0.36885051, -0.39111768,  0.01860959,  0.58717613,  0.08037357,
        1.27226064, -0.66023757,  0.65523138,  0.01480391,  0.03709066,
        0.04716515,  1.01057874, -0.17887257,  0.02529794, -0.00

#### Parameters using iterative method ( Gradient Descent iterative method)

In [62]:
gradient_descent(df3,1000,0.000001)

(0.15989214174067923,
 array([ 0.09250802, -2.10562034, -0.19641832, -0.35205565,  0.28577124,
         0.0957374 ,  0.54560709,  0.69164503, -0.18541106, -0.21201196,
        -0.67311597, -0.19285129,  0.44192121,  1.15708132, -0.63073413,
         2.14380163,  0.01553301,  0.13200608,  0.02380999, -0.75300981,
         0.59511108, -0.34153763, -0.03450405,  0.08170065, -0.29960883,
         0.09144054, -1.1425564 ,  0.10036185,  0.03074346,  0.03097076,
         0.34442867,  0.91725474, -0.10884724,  0.62390082,  0.056974  ,
         0.56662107, -0.17540709,  0.39468878,  0.02900727,  0.26265882,
        -0.1301798 ,  0.37679192,  0.23210444,  0.47605074,  0.56612776,
         0.06664002,  0.18331375,  0.17999684,  0.15940563, -0.39419423,
         0.27575858, -0.20303336,  0.05957212,  2.2424814 , -0.57989419,
        -0.36732782, -0.38970175,  0.0176897 ,  0.58595046,  0.07923003,
         1.27186819, -0.66053936,  0.65584372,  0.01515247,  0.03904347,
         0.04919205,  1.01022

#### (b) Compare the time taken and space requirements of two approaches.


#### Time Taken by direct approach

In [63]:
import timeit
%timeit [[direct_method(df3)]]

4.62 ms ± 75.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


#### Time Taken by iterative  approach

In [68]:
#import timeit
%timeit [gradient_descent(df3,100,0.000001)]

10.3 ms ± 122 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


#### (c) Describe the search space and cross validation techniques used to find the values of the hyperparameters of the model.


In [159]:
len(df3)

800

In [162]:
len(df3.columns)

201


####  Data Set 3 :
####  Cardinality or Data Set length : N = 800
####  Number of attributes :  n = 200
####  Dimensions :       x = N*(n+1) = 800 * 201        ;       w= (n+1) * 1 = 201 * 1       ;        y_pred = N * 1 = 800 * 1

####  K_Fold Cross Validation Technique is used with Number of Folds k = 5
####  From the given data set length = 200 
####  80% of data is used to Train the model
####  20% of data is used to Test the model

####  TRAIN THE MODEL :
#### Linear Regression is Performed by using Iterative techniques called Batch Gradient Descent from the 80%  Data = Train Data=640
####  For Each Fold i,e 1 to k (i,e 5 )  The Parameters W and W0 are stored in a list  len(W) = len(W0) = 5
####  Now we found Parameters (W,W0) in each fold , which will be used in testing the 20% Data

#### TESTING THE MODEL :
#### WIth available 20% Data = Test Data = 160
#### Find y_predicted i,e y_pred = (w)Transpose * x + w0
#### As length of x =160 ,SO  lenght of Y_pred = 160
#### Now Compare this Y_pred with Y_test
####  So that we can see the performance of the model
####  How well the Model is performing For the Test Data With which the Model hasn't Trained

#### The Output Results of both Y_pred and Y_train are given in the Output of the Program to compare

In [69]:
KFold_cross_validation_train(df3_normalized,1000,0.00000001,5)

 Data Set Lenght : 800 
640
640
w0:0.0024822661967478865 w:[0.00125512 0.00114372 0.00129777 0.00119324 0.00132143 0.00129924
 0.00122818 0.00121621 0.00122408 0.00122062 0.00118189 0.00123767
 0.00127062 0.00127938 0.00123679 0.00132011 0.00123289 0.00121174
 0.00123363 0.00117413 0.00126007 0.00121848 0.00120139 0.0012925
 0.00124615 0.0012719  0.00115487 0.00125415 0.00119282 0.00123791
 0.00125195 0.00122083 0.00125535 0.00125866 0.00123948 0.00122376
 0.00126181 0.00124605 0.00125099 0.00127643 0.00120456 0.00124516
 0.00122207 0.00122742 0.00123903 0.00124795 0.00122742 0.00125681
 0.0012452  0.00120339 0.00124954 0.00122364 0.00127655 0.00132423
 0.00124051 0.00124038 0.00122576 0.00121742 0.00124399 0.00122267
 0.00125373 0.00118407 0.00129114 0.00124419 0.00126319 0.00127822
 0.00128351 0.0012338  0.00125467 0.0012177  0.0011646  0.00127205
 0.00115854 0.00128993 0.00125686 0.00124945 0.00124893 0.00118815
 0.00124045 0.00126308 0.00121232 0.00128804 0.00123516 0.00122179
 0.0

w0:0.00246832185309439 w:[0.0012272  0.00117976 0.00124734 0.00117856 0.00132312 0.0012613
 0.00123913 0.00121483 0.00122062 0.00119887 0.00117334 0.00123561
 0.00127206 0.00126758 0.00121913 0.00131323 0.00124622 0.00119985
 0.00121549 0.00116748 0.00128439 0.0012499  0.00120788 0.00126195
 0.00124163 0.0012728  0.00118019 0.00130588 0.00121254 0.00121899
 0.00126928 0.00123362 0.00122574 0.00123179 0.00125037 0.00123475
 0.00122674 0.00119888 0.00124712 0.00127088 0.00125452 0.00122805
 0.00122882 0.0012208  0.00123913 0.00126629 0.00124981 0.00122569
 0.00124242 0.00120726 0.00126507 0.0012265  0.00125989 0.00129487
 0.00124185 0.00124014 0.00122931 0.0011898  0.00121034 0.00123658
 0.00123621 0.00118722 0.00126913 0.00122045 0.00121394 0.00129716
 0.0012962  0.00121276 0.0012463  0.00119718 0.00116154 0.00124649
 0.00115045 0.00128231 0.00125636 0.0012514  0.00123513 0.00117074
 0.00120886 0.00126118 0.00118129 0.00128622 0.00123129 0.00122603
 0.00120542 0.00122834 0.00119167 0.00

In [70]:
KFold_cross_validation_test(df3_normalized,1000,0.00000001,5)

 Data Set Lenght : 800 
 
Lenth of x_train:160
Lenth of y_train:160
 
For Kth Fold: 0 
 
 
 y_pred:0.6253578436507653 y_train:0.417103912196849
 
 y_pred:0.6202069036168106 y_train:0.569130819614091
 
 y_pred:0.6204068011384404 y_train:0.40701150646132056
 
 y_pred:0.6087801139010087 y_train:0.40572242874845105
 
 y_pred:0.6259399570969133 y_train:0.46123278456363953
 
 y_pred:0.6178579357078738 y_train:0.5805735528412109
 
 y_pred:0.6166799054368011 y_train:0.544326429456541
 
 y_pred:0.6174509711678758 y_train:0.5794052044609665
 
 y_pred:0.6224773994490058 y_train:0.20010975393875025
 
 y_pred:0.6223375630551525 y_train:0.35731279872543814
 
 y_pred:0.6190610290647566 y_train:0.4250217737652682
 
 y_pred:0.6126711376197602 y_train:0.38982474774296333
 
 y_pred:0.6262361046036392 y_train:0.5230801911842804
 
 y_pred:0.6191284502339267 y_train:0.3648150115064613
 
 y_pred:0.6231529747855579 y_train:0.423349619401664
 
 y_pred:0.6197960589225011 y_train:0.5812320764737121
 
 y_pred:0.6

 
For Kth Fold: 2 
 
 
 y_pred:0.6128880382936871 y_train:0.2415648787395999
 
 y_pred:0.6080941000310373 y_train:0.5435262878385555
 
 y_pred:0.6087578388740628 y_train:0.351414409630023
 
 y_pred:0.6158522514324847 y_train:0.3735988670561161
 
 y_pred:0.6130754481936113 y_train:0.3273959992919101
 
 y_pred:0.6133571129329132 y_train:0.4511959638874137
 
 y_pred:0.6208142307106405 y_train:0.666372809346787
 
 y_pred:0.602741653863138 y_train:0.27969198088157193
 
 y_pred:0.6099694398604584 y_train:0.4772596919808816
 
 y_pred:0.6197756779353991 y_train:0.21979819437068507
 
 y_pred:0.6163643264152405 y_train:0.5911170118605064
 
 y_pred:0.6271671819284965 y_train:0.4266029385731988
 
 y_pred:0.6155324436510646 y_train:0.29370862099486633
 
 y_pred:0.6120489485026898 y_train:0.45770755885997516
 
 y_pred:0.6109039341040271 y_train:0.6239723844928305
 
 y_pred:0.6132299092180561 y_train:0.516317932377412
 
 y_pred:0.6192059748470553 y_train:0.5095592140201806
 
 y_pred:0.606953463613710

### (d) Report the values of the hyperparameters.


### Hyperparameters Tried For a range between 10^-2 to 10^-8

### (e) Report the parameters of the model.



In [482]:
gradient_descent(df3,1000000,0.00000001)

(1.4983434049654,
 array([ 0.09046386, -2.11020815, -0.20125642, -0.35427716,  0.2841442 ,
         0.09324901,  0.54461284,  0.69207336, -0.18959324, -0.2121276 ,
        -0.67664885, -0.19572388,  0.44025743,  1.15663165, -0.6332092 ,
         2.14370642,  0.01374135,  0.13007226,  0.02107402, -0.75211778,
         0.59323248, -0.34492612, -0.03625814,  0.0799886 , -0.29763116,
         0.08737728, -1.14745302,  0.09735791,  0.0284868 ,  0.0287256 ,
         0.34334385,  0.91398852, -0.11053087,  0.62313825,  0.05429628,
         0.5608058 , -0.17872754,  0.39277941,  0.02658697,  0.25990142,
        -0.12645259,  0.37581969,  0.231598  ,  0.47376577,  0.56327154,
         0.06547099,  0.18015373,  0.18037424,  0.15650916, -0.39946012,
         0.27196158, -0.20294804,  0.05561844,  2.24546674, -0.58338921,
        -0.37067133, -0.39126848,  0.0139055 ,  0.585795  ,  0.07610891,
         1.27002541, -0.6612574 ,  0.65235348,  0.01042921,  0.03231044,
         0.04392801,  1.00646147,